### PR TITLE
Add AlloyDB user support

### PR DESCRIPTION
**Pull request by DanielRieske**
If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [x] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
google_alloydb_user
```

Closes: https://github.com/hashicorp/terraform-provider-google/issues/14272

### CODE REVIEW

**Comment by DanielRieske** [comment]
Added tests to increase test coverage for the missing field, can we run the tests again?

**Comment by DanielRieske** [comment]
Anyway of getting the log where the test `TestAccAlloydbUser_alloydbUserExample` went wrong?

I tested them on my own account on which it passed.

```

terraform-provider-google git:(main) ✗ make testacc TEST=./google TESTARGS='-run=TestAccAlloydbUser_alloydbUserExample'
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccAlloydbUser_alloydbUserExample -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccAlloydbUser_alloydbUserExample
=== PAUSE TestAccAlloydbUser_alloydbUserExample
=== CONT  TestAccAlloydbUser_alloydbUserExample
--- PASS: TestAccAlloydbUser_alloydbUserExample (909.38s)
PASS
ok      github.com/hashicorp/terraform-provider-google/google   910.034s

```

**Comment by DanielRieske** [comment]
I realised that I never ran the tests at the same time, by doing this I realised that the `compute_networks` being created didn't have a unique name and therefore failing.

```
terraform-provider-google git:(main) ✗ make testacc TEST=./google TESTARGS='-run=TestAccAlloydbUser_'
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccAlloydbUser_ -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccAlloydbUser_alloydbUserExample
=== PAUSE TestAccAlloydbUser_alloydbUserExample
=== RUN   TestAccAlloydbUser_alloydbUserIamUserExample
=== PAUSE TestAccAlloydbUser_alloydbUserIamUserExample
=== CONT  TestAccAlloydbUser_alloydbUserExample
=== CONT  TestAccAlloydbUser_alloydbUserIamUserExample
--- PASS: TestAccAlloydbUser_alloydbUserExample (1136.85s)
--- PASS: TestAccAlloydbUser_alloydbUserIamUserExample (1171.33s)
PASS
ok      github.com/hashicorp/terraform-provider-google/google   1171.995s
```

File: mmv1/templates/terraform/examples/alloydb_user_iam_user.tf.erb
Comment (slevenick) [inline mmv1/templates/terraform/examples/alloydb_user_iam_user.tf.erb:15]: This should change to a parameterized value so it gets a random suffix attached. Right now this is causing the failure because it collides with another network named the same

**Review by slevenick — CHANGES_REQUESTED** [review]
Error is: global/networks/alloydb-cluster' already exists, alreadyExists


Please change the name to be parameterized so it gets randomized

File: mmv1/templates/terraform/examples/alloydb_user_iam_user.tf.erb
Comment (DanielRieske) [reply mmv1/templates/terraform/examples/alloydb_user_iam_user.tf.erb:15]: @slevenick I am skipping the generation of tests because this particular resource does not work well with the autogenerated tests.

If we destroy the cluster or instance during the CheckDestroy phase that will result in Internal 500 errors when trying to verify if an user still exists, therefore I need to check it while the parent resource still exists.

**Comment by DanielRieske** [comment]
This was changed in the "handwritten" tests and should work.

File: mmv1/templates/terraform/examples/alloydb_user_iam_user.tf.erb
Comment (slevenick) [reply mmv1/templates/terraform/examples/alloydb_user_iam_user.tf.erb:15]: That's.... frightening. What do we expect the user experience to be when they created a user within Terraform and then they destroyed it via another tool? They would hit the 500 and be confused about what is happening

Comment (slevenick) [reply mmv1/templates/terraform/examples/alloydb_user_iam_user.tf.erb:15]: Can you share either the error message in logs or a use case that triggers it? I can reach out to the internal team because returning a 500 on an attempted GET is pretty bad.

Comment (DanielRieske) [reply mmv1/templates/terraform/examples/alloydb_user_iam_user.tf.erb:15]: @slevenick I did not realise the impact in certain scenario's on users and didn't report it, for future contributions I will keep this in mind.

There are a few scenario's where this can happens but usually boils down to not having an instance in it's cluster.

I tested the following 2 scenario's:

_**You create a cluster and try to add an user without creating an instance.**_

```

2023-07-26T20:57:40.102+0200 [DEBUG] provider.terraform-provider-google: ---[ REQUEST ]---------------------------------------
2023-07-26T20:57:40.102+0200 [DEBUG] provider.terraform-provider-google: POST /v1/projects/playground-x/locations/europe-west4/clusters/tf-test-alloydb-cluster/users?alt=json&userId=test HTTP/1.1
2023-07-26T20:57:40.102+0200 [DEBUG] provider.terraform-provider-google: Host: alloydb.googleapis.com
2023-07-26T20:57:40.102+0200 [DEBUG] provider.terraform-provider-google: User-Agent: Terraform/1.5.0 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google/dev
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google: Content-Length: 67
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google: Content-Type: application/json
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google: Accept-Encoding: gzip
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google: 
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google: {
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google:  "databaseRoles": [
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google:   "alloydbiamuser"
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google:  ],
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google:  "userType": "ALLOYDB_IAM_USER"
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google: }
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google: 
2023-07-26T20:57:40.103+0200 [DEBUG] provider.terraform-provider-google: -----------------------------------------------------
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: 2023/07/26 20:57:40 [DEBUG] Google API Response Details:
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: ---[ RESPONSE ]--------------------------------------
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: HTTP/2.0 500 Internal Server Error
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: Cache-Control: private
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: Content-Type: application/json; charset=UTF-8
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: Date: Wed, 26 Jul 2023 18:57:40 GMT
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: Server: ESF
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: Vary: Origin
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: Vary: X-Origin
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: Vary: Referer
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: X-Content-Type-Options: nosniff
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: X-Frame-Options: SAMEORIGIN
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: X-Xss-Protection: 0
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: 
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google: {
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google:   "error": {
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google:     "code": 500,
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google:     "message": "An internal error has occurred (4e2d52b2-e97b-466a-9d22-a1a2c9b499bc)",
2023-07-26T20:57:40.263+0200 [DEBUG] provider.terraform-provider-google:     "status": "INTERNAL"
2023-07-26T20:57:40.264+0200 [DEBUG] provider.terraform-provider-google:   }
2023-07-26T20:57:40.264+0200 [DEBUG] provider.terraform-provider-google: }

```

```
│ Error: Error creating User: timeout while waiting for state to become 'success' (timeout: 1m0s)
│ 
│   with google_alloydb_user.default,
│   on main.tf line 61, in resource "google_alloydb_user" "user":
│   34: resource "google_alloydb_user" "user" {
```


_**You create a cluster, instance & user through Terraform and we delete the instance and try to pull the state of the user resource**_

```

2023-07-26T20:23:52.799+0200 [DEBUG] provider.terraform-provider-google: ---[ REQUEST ]---------------------------------------
2023-07-26T20:23:52.799+0200 [DEBUG] provider.terraform-provider-google: GET /v1/projects/playground-x/locations/europe-west4/clusters/tf-test-alloydb-cluster/users/test?alt=json HTTP/1.1
2023-07-26T20:23:52.799+0200 [DEBUG] provider.terraform-provider-google: Host: alloydb.googleapis.com
2023-07-26T20:23:52.799+0200 [DEBUG] provider.terraform-provider-google: User-Agent: Terraform/1.5.0 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google/dev
2023-07-26T20:23:52.799+0200 [DEBUG] provider.terraform-provider-google: Content-Type: application/json
2023-07-26T20:23:52.799+0200 [DEBUG] provider.terraform-provider-google: Accept-Encoding: gzip
2023-07-26T20:23:52.799+0200 [DEBUG] provider.terraform-provider-google: 
2023-07-26T20:23:52.799+0200 [DEBUG] provider.terraform-provider-google: 
2023-07-26T20:23:52.800+0200 [DEBUG] provider.terraform-provider-google: -----------------------------------------------------
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: 2023/07/26 20:23:52 [DEBUG] Google API Response Details:
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: ---[ RESPONSE ]--------------------------------------
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: HTTP/2.0 500 Internal Server Error
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: Cache-Control: private
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: Content-Type: application/json; charset=UTF-8
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: Date: Wed, 26 Jul 2023 18:23:52 GMT
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: Server: ESF
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: Vary: Origin
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: Vary: X-Origin
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: Vary: Referer
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: X-Content-Type-Options: nosniff
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: X-Frame-Options: SAMEORIGIN
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: X-Xss-Protection: 0
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: 
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: {
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google:   "error": {
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google:     "code": 500,
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google:     "message": "An internal error has occurred (044b08c2-93bd-4981-a397-05c5c1b32f2b)",
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google:     "status": "INTERNAL"
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google:   }
2023-07-26T20:23:52.837+0200 [DEBUG] provider.terraform-provider-google: }

```


```
│ Error: Error when reading or editing AlloydbUser "projects/playground-x/locations/europe-west4/clusters/tf-test-alloydb-cluster/users/test": googleapi: Error 500: An internal error has occurred (570d762a-8316-4854-b15c-a40116fb4e97)
│ 
│   with google_alloydb_user.user,
│   on main.tf line 61, in resource "google_alloydb_user" "user":
│   61: resource "google_alloydb_user" "user" {

```


While I was at it I tested what happens with an invalid cluster name and it got rejected as a 404.

Given these results I am unsure if this should be merged at this time.

**Comment by slevenick** [comment]
@GauravJain21 can you take a look at the error message here? This looks to be a bug on the service side, we shouldn't be getting a 500 response

File: mmv1/products/alloydb/User.yaml
Comment (slevenick) [inline mmv1/products/alloydb/User.yaml:19]: Given that you've set up the update support here can you add a test that updates all possible fields within the resource? That will help us validate both that the update support works and that all fields marked as updatable actually are updatable.

File: mmv1/products/alloydb/User.yaml
Comment (slevenick) [inline mmv1/products/alloydb/User.yaml:70]: Is it preferable to have this set up as a single parent resource rather than a series of project/location/cluster?

Generally we have each broken out into an individual field.

cc @GauravJain21

File: mmv1/products/alloydb/User.yaml
Comment (slevenick) [inline mmv1/products/alloydb/User.yaml:101]: Generally we don't need the UNSPECIFIED enum value as it isn't actually settable. So probably remove this.

File: mmv1/templates/terraform/examples/alloydb_user_iam_user.tf.erb
Comment (slevenick) [reply mmv1/templates/terraform/examples/alloydb_user_iam_user.tf.erb:15]: @GauravJain21 please take a look at this

**Review by slevenick — CHANGES_REQUESTED** [review]
Needs some changes before merging, but the main questions come around the stability of this API for integration, especially around the 500 error returned in some cases.

At this point we should wait for @GauravJain21 to review

File: mmv1/products/alloydb/User.yaml
Comment (DanielRieske) [reply mmv1/products/alloydb/User.yaml:70]: That would mean that the user has to set it's location and I'd argue this resource isn't a regional object on it's own but rather part of a regional object.

As the cluster dictates what region the user resides in I wouldn't give the user the option to specify location.

Comment (slevenick) [reply mmv1/products/alloydb/User.yaml:70]: True. On the other hand existing resources like `google_sql_user` work this way, where the project and instance are both specified independently: https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/sql_user#project

File: mmv1/products/alloydb/User.yaml
Comment (DanielRieske) [reply mmv1/products/alloydb/User.yaml:19]: I added an additional step to the testplan that updates all possible fields.

File: mmv1/products/alloydb/User.yaml
Comment (DanielRieske) [reply mmv1/products/alloydb/User.yaml]: Removed the `UNSPECIFIED` value.

File: mmv1/products/alloydb/User.yaml
Comment (DanielRieske) [reply mmv1/products/alloydb/User.yaml:71]: The biggest difference here is that `google_sql_user` doesn't rely on a region in his end-point nor is it a field that can be set.
It is implied that it lives in the region where the instance resides in.

`POST https://sqladmin.googleapis.com/v1/projects/{project}/instances/{instance}/users`

I am curious what others believe the right approach is here.

**Comment by DanielRieske** [comment]
`TestAccAlloydbUser_alloydbUser` failed because of an incorrect cluster reference, I corrected the mistake and ran the tests on my own account.

```
➜  terraform-provider-google git:(main) ✗ make testacc TEST=./google TESTARGS='-run=TestAccAlloydbUser_'         
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccAlloydbUser_ -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccAlloydbUser_alloydbUser
=== PAUSE TestAccAlloydbUser_alloydbUser
=== RUN   TestAccAlloydbUser_alloydbUserIamUser
=== PAUSE TestAccAlloydbUser_alloydbUserIamUser
=== CONT  TestAccAlloydbUser_alloydbUser
=== CONT  TestAccAlloydbUser_alloydbUserIamUser
--- PASS: TestAccAlloydbUser_alloydbUserIamUser (1173.06s)
--- PASS: TestAccAlloydbUser_alloydbUser (1232.61s)
PASS
ok      github.com/hashicorp/terraform-provider-google/google   1233.281s
```


Anyway we can re-run the tests?

**Comment by slevenick** [comment]
Tests look good, I'm going to hold off on more review until @GauravJain21 has a chance to look at this next week.

**Comment by DanielRieske** [comment]
Hi @GauravJain21, did you have an opportunity yet to take a look at this?

**Comment by GauravJain21** [comment]
Hey Daniel,
I am currently working on something else. I intend to review this next week
or the week after.
Best,
Gaurav Jain

**Comment by DanielRieske** [comment]
Hi @GauravJain21, anything I can do to help at this point?

**Comment by GauravJain21** [comment]
Hey Daniel,
We have stopped taking in PR requests for major features from the external community contribution. An internal developer is working for adding Users API support to Terraform.
Best,
Gaurav

**Comment by DanielRieske** [comment]
@GauravJain21 Thank you for your response and sad to hear that major features created by the external community won't be accepted anymore.

Will there be an official statement on what type of contributions will or will not be considered? I truly enjoy working on this product and have thoroughly enjoyed contributing to it.

But if certain contributions won't be accepted anymore then I'd like to what extent so I won’t spend time on contributions that might not get accepted.

Thanks a lot!

**Comment by alexgartner-bc** [comment]
@GauravJain21 any update on this?

**Comment by GauravJain21** [comment]
We are tracking it [here](https://github.com/hashicorp/terraform-provider-google/issues/14272). Can you close this as duplicate. [PR](https://github.com/GoogleCloudPlatform/magic-modules/pull/8920) has been raised.

**Pull request closed without merging**
